### PR TITLE
Add TERM iff TERM not defined in container when podman exec -t

### PR DESCRIPTION
**Pull request by rhatdan**
Fixes: https://github.com/containers/podman/issues/20334

#### Does this PR introduce a user-facing change?

```release-note
Podman exec -t will define TERM environment variable even if container is not running with a terminal.
```

### CODE REVIEW

**Comment by rhatdan** [comment]
@edsantiago Not sure what is going on, but the out put of env from inside of container seems to have a ^M associated with it, not sure how to remove it nicely.

File: libpod/container.go
Comment (vrothberg) [inline libpod/container.go:716]: Please add a doc comment.

File: pkg/util/utils.go
Comment (vrothberg) [inline pkg/util/utils.go:1223]: Since there's only one caller and no uni tests, can we move it directly to `pkg/domain/infra/abi/containers.go`?

File: test/system/075-exec.bats
Comment (vrothberg) [inline test/system/075-exec.bats:172]: The `printenv echo $TERM` confuses me.  Do you want to `echo` _or_ `printenv`?

File: test/system/075-exec.bats
Comment (vrothberg) [inline test/system/075-exec.bats:184]: This entire block looks a duplicate of the previous one.

File: pkg/api/handlers/compat/exec.go
Comment (Luap99) [inline pkg/api/handlers/compat/exec.go:62]: I don't see why this should ever depend on `ctr.Terminal()`? Either exec is started with tty or not, I don't see why the container option matters in this case.

File: test/system/075-exec.bats
Comment (edsantiago) [reply test/system/075-exec.bats:172]: Not only that, the `echo $TERM` (without single quotes) will interpret `$TERM` in the context of the running shell (BATS), not the container. This should be `printenv TERM`, as it is below.

**Review by edsantiago — COMMENTED** [review]
`^M` is the result of `-t`. For a tighter test, I recommend:
```bash
    cr=$'\r'
    ...
    assert "$output" = "whatever$cr" "..."
```

File: pkg/util/utils.go
Comment (rhatdan) [reply pkg/util/utils.go:1223]: There are two callers, one from api/handlers an one from domain/infra/abi

File: test/system/075-exec.bats
Comment (rhatdan) [reply test/system/075-exec.bats:172]: Yes this was a leak, attempting to get this, should all be printenv TERM.

File: test/system/075-exec.bats
Comment (edsantiago) [reply test/system/075-exec.bats:184]: Way too many confusing combinations. Might I make a suggestion?
```bash
@test "podman exec --tty" {
    # Outer loops: different variations on the RUN container
    for run_opt_t in "" "-t"; do
        for run_term_env in "" "explicit_RUN_term"; do
            local run_opt_env=
            if [[ -n "$run_term_env" ]]; then
                run_opt_env="--env=TERM=$run_term_env"
            fi
            run_podman run -d $run_opt_t $run_opt_env --name test $IMAGE top

            # Inner loops: different variations on EXEC
            for exec_opt_t in "" "-t"; do
                for exec_term_env in "" "explicit_EXEC_term"; do
                    # What to expect.
                    local expected=
                    # "exec -t" defaults to current $TERM
                    if [[ -n "$exec_opt_t" ]]; then
                        expected="$TERM"
                    fi
                    # ...unless "run -t", which forces xterm (always xterm)
                    if [[ -n "$run_opt_t" ]]; then
                        expected="xterm"
                    fi
                    # ...unless overridden by explicit --env
                    if [[ -n "$run_term_env$exec_term_env" ]]; then
                        # (exec overrides run)
                        expected="${exec_term_env:-$run_term_env}"
                    fi

                    local exec_opt_env=
                    if [[ -n "$exec_term_env" ]]; then
                        exec_opt_env="--env=TERM=$exec_term_env"
                    fi

                    run_podman exec $exec_opt_t $exec_opt_env test sh -c 'echo -n $TERM'
                    assert "$output" = "$expected" "run $run_opt_t $run_opt_env, exec $exec_opt_t $exec_opt_env"
                done
            done

            run_podman rm -f -t0 test
        done
    done
```
This very clearly demonstrates what I consider a confusing aspect of `run -t` with `exec -t`. If it's deliberate, fine, but to me it violates POLA that `exec -t` depends on whether the container was `run` with/without `-t`.

Comment (rhatdan) [reply test/system/075-exec.bats]: This is fine, it does loose the test to make sure TERM is not set in the non `run -t` and non `exec -t` case, since theoretically TERM="" versus TERM not set. But really not important.

Comment (edsantiago) [reply test/system/075-exec.bats]: That would be fairly easy to fix:
```bash
    local expect_rc=0
    if [[ -z "$expect_output" ]]; then
         expect_rc=1
    fi
    ...
    run_podman $expect_rc exec [...] printenv TERM
```
...but that then requires defining `cr=$'\r'` and requiring it in the `-t` cases. Which I think might be fine, but it makes the test more complicated, and the loop is already too unwieldy. LMK if you'd like the full `\r` set of changes.

**Comment by rhatdan** [comment]
@edsantiago it is blowing up with an unexpected terminal name `dumb`?

**Comment by edsantiago** [comment]
Argh! I can reproduce via `env -u TERM hack/bats --rootless 075:tty </dev/null`. (I had only tested via `TERM= hack/bats`, which is obviously not the same). Now looking for a fix.

**Comment by edsantiago** [comment]
This fixes the tests..... but only for local podman. For remote, `podman run (no opts)` with `exec -t` gets the podman-system-service `$TERM`, not the invoking client's. My gut tells me that's wrong, but it's too late in my day right now for me to fully think it through.
```diff
diff --git a/test/system/075-exec.bats b/test/system/075-exec.bats
index ec7d59536..324c9bbfe 100644
--- a/test/system/075-exec.bats
+++ b/test/system/075-exec.bats
@@ -177,3 +177,3 @@ load helpers
             fi
-            run_podman run -d $run_opt_t $run_opt_env --name test $IMAGE top
+            TERM=run-term run_podman run -d $run_opt_t $run_opt_env --name test $IMAGE top
 
@@ -184,5 +184,5 @@ load helpers
                     local expected=
-                    # "exec -t" defaults to current $TERM
+                    # "exec -t" defaults to the process TERM
                     if [[ -n "$exec_opt_t" ]]; then
-                        expected="$TERM"
+                        expected="exec-term"
                     fi
@@ -203,3 +203,3 @@ load helpers
 
-                    run_podman exec $exec_opt_t $exec_opt_env test sh -c 'echo -n $TERM'
+                    TERM=exec-term run_podman exec $exec_opt_t $exec_opt_env test sh -c 'echo -n $TERM'
                     assert "$output" = "$expected" "run $run_opt_t $run_opt_env, exec $exec_opt_t $exec_opt_env"
```
Note that this does not test the `env -u TERM run_podman` case, because we can't do that. Can't do `env` with `run_podman`. This keep getting trickier. I'm sorry to have introduced so much hassle, but am slightly glad that we're giving attention to the corner cases.

**Comment by edsantiago** [comment]
@rhatdan this is going to fail. I guess I wasn't clear in my comments, sorry.

The podman-remote case uses `$TERM` on the server. I think this is wrong:
```console
$ TERM=sdfsdf bin/podman system service --timeout=0
```
```
$ bin/podman-remote run -d --name foo quay.io/libpod/testimage:20221018 top
<sha>
$ bin/podman-remote exec -t foo printenv TERM
sdfsdf
```
Note that `podman-remote run -t` shows `xterm`, not `sdf`.

Anyhow, tests will fail until this is resolved. Either skip part of the tests under remote, or fix the underlying code, and I would prefer to fix the underlying code. I think `podman-remote exec` should use client TERM, not server. But I may be wrong.

**Comment by rhatdan** [comment]
@edsantiago I changed the `podman run -t` behavior to do a Getenv("TERM") before failing over to "xterm"
Now it should work on --remote.

**Comment by edsantiago** [comment]
I'm not sure this is better? It still fails remote (no change), and now I think the `exec` behavior is worse: `exec -t`, on a container with `run -t`, now gets the **run** TERM, not the exec one. I think that's a step in the opposite direction than we want?

Anyhow, this is going to fail all system tests, not just remote.

**Comment by edsantiago** [comment]
My bad. This is actually a very small net gain (`run -t` now honors TERM) but I still think that `exec -t` should win. This patch makes it work under podman-local:
```diff
diff --git a/test/system/075-exec.bats b/test/system/075-exec.bats
index 324c9bbfe..14f965063 100644
--- a/test/system/075-exec.bats
+++ b/test/system/075-exec.bats
@@ -188,5 +188,5 @@ load helpers
                     fi
-                    # ...unless "run -t", which forces xterm (always xterm)
+                    # ...unless "run -t", which forces the run-process TERM
                     if [[ -n "$run_opt_t" ]]; then
-                        expected="xterm"
+                        expected="run-term"
                     fi
```
...but remote is still 100% broken and I still lean toward thinking that it should be fixed in the code, not the tests.

What I _can_ do, though, is kludge up the tests so they pass under remote, then sweep the podman-remote-exec problem under the rug for another bug week. Should I do that?

**Comment by edsantiago** [comment]
copy-paste this, it will work under local and remote. I've made the warning messages very obnoxious as an incentive to getting that bug fixed (or at least getting the issue resolved one way or another).
```bash
@test "podman exec --tty" {
    # Outer loops: different variations on the RUN container
    for run_opt_t in "" "-t"; do
        for run_term_env in "" "explicit_RUN_term"; do
            local run_opt_env=
            if [[ -n "$run_term_env" ]]; then
                run_opt_env="--env=TERM=$run_term_env"
            fi
            TERM=run-term run_podman run -d $run_opt_t $run_opt_env --name test $IMAGE top

            # Inner loops: different variations on EXEC
            for exec_opt_t in "" "-t"; do
                for exec_term_env in "" "explicit_EXEC_term"; do
                    # What to expect.
                    local expected=
                    # "exec -t" defaults to the process TERM
                    if [[ -n "$exec_opt_t" ]]; then
                        expected="exec-term"
                    fi
                    # ...unless "run -t", which forces the run-process TERM
                    if [[ -n "$run_opt_t" ]]; then
                        expected="run-term"
                    fi
                    # ...unless overridden by explicit --env
                    if [[ -n "$run_term_env$exec_term_env" ]]; then
                        # (exec overrides run)
                        expected="${exec_term_env:-$run_term_env}"
                    fi

                    local exec_opt_env=
                    if [[ -n "$exec_term_env" ]]; then
                        exec_opt_env="--env=TERM=$exec_term_env"
                    fi

                    local desc="run $run_opt_t $run_opt_env, exec $exec_opt_t $exec_opt_env"
                    TERM=exec-term run_podman exec $exec_opt_t $exec_opt_env test sh -c 'echo -n $TERM'
                    # FIXME FIXME FIXME! #xxxxx: podman-remote run/exec -t use server TERM, not client
                    if is_remote && [[ -n "$run_opt_t$exec_opt_t" ]] && [[ -z "$run_term_env$exec_term_env" ]]; then
                        echo "# [[ FIXME: podman-remote -t uses server TERM, not client : $desc ]]" >&3
                    else
                        # The normal case
                        assert "$output" = "$expected" "$desc"
                    fi
                done
            done

            run_podman rm -f -t0 test
        done
    done
}
```

**Comment by rhatdan** [comment]
Switching to TERM on the client side is a potential breaking change, but might make the most sense.

Up til now, exec used TERM from the Container, the change that triggered this PR was that podman run --tty=false, no longer sets a TERM environment variable.  This was causing podman exec --tty=true to said terminals to not work correctly without a TERM set.

This PR still takes the TERM from the container if it is set, and then takes if from the client, when it is not set.

Changing to always take TERM from the client for podman exec, might make the most sense, although in the server case, you could have a TERM on the client side that is not understood on the server side.  Perhaps a TERM from a Mac or Windows is not understood in linux.  (The current code is susceptible to the same issue).

@vrothberg @Luap99 @edsantiago @baude Thoughts?

**Review by vrothberg — APPROVED** [review]
I am OK with the change
LGTM

**Comment by Luap99** [comment]
Also worth pointing out before committing to such changes we should verify what docker does. If docker doesn't read from the environment neither should we IMO.

**Comment by edsantiago** [comment]
>Why are we looking up the TERM from the host environment exactly?

My fault: I got sick of all the churn on this, and added tests, and my new test found what I consider to be weird behavior. But yeah, I guess we should figure out what docker does and copy it.

**Comment by rhatdan** [comment]
Basically docker sets the TERM=xterm if it not defined in the container image on docker run.
Docker sets the TERM=xterm if TERM is not defined in the container on docker exec.

**Comment by edsantiago** [comment]
[exec something leak](https://api.cirrus-ci.com/v1/artifact/task/6610353904156672/html/sys-remote-rawhide-root-host-sqlite.log.html#t--00233), @giuseppe PTAL

File: test/system/075-exec.bats
Comment (edsantiago) [inline test/system/075-exec.bats:184]: Just "xterm", not "the process xterm". (You global-replaced TERM, that yielded some nonsense results)

File: test/system/075-exec.bats
Comment (edsantiago) [inline test/system/075-exec.bats:188]: here too

File: test/system/075-exec.bats
Comment (edsantiago) [inline test/system/075-exec.bats:205]: No need for this comment any more, please remove

File: test/system/075-exec.bats
Comment (edsantiago) [reply test/system/075-exec.bats:188]: Actually, this block is now absurd. It should be collapsed, both `if`s should become:
```bash
    # if -t is set anywhere, either run or exec, go with xterm
    if [[ -n "$run_opt_t$exec_opt_t" ]]; then
```

**Review by edsantiago — CHANGES_REQUESTED** [review]
You have to repush (once #20404 merges) to fix CI anyway, so, some fix requests

**Comment by edsantiago** [comment]
20404 (CI fix) is merged. You can now rebase. Please do fix the issues I reported above.

**Comment by edsantiago** [comment]
Flake is new to me. Restarted.

/lgtm
/hold

This has been quite the ordeal. Thank you @rhatdan for your patience with my obnoxious insistence on testing.

**Pull request merged**
